app/src/main/java/org/mozilla/tv/firefox/navigationoverlay/ToolbarUiController.kt

```diff
@@ -36,17 +45,40 @@ class ToolbarUiController(
 
     private var hasUserChangedURLSinceEditTextFocused = false
 
+    private lateinit var tooltip: PopupWindow
+    private lateinit var tooltipView: View
+
     fun onCreateView(layout: View, viewLifecycleOwner: LifecycleOwner, fragmentManager: FragmentManager) {
         val toolbarClickListener = ToolbarOnClickListener()
         layout.topNavContainer.forEach {
             it.nextFocusDownId = layout.navUrlInput.id
             if (it.isFocusable) it.setOnClickListener(toolbarClickListener)
+
+            it.setOnFocusChangeListener { _, hasFocus ->
+                if (hasFocus) showTooltip(it)
+                else tooltip.dismiss() // Hide the tooltip when the button is not focused
+            }
         }
 
+        val layoutInflater = layout.context.systemService<LayoutInflater>(Context.LAYOUT_INFLATER_SERVICE)
+        tooltipView = layoutInflater.inflate(R.layout.tooltip, null)
+        tooltip = PopupWindow(tooltipView, WRAP_CONTENT, WRAP_CONTENT, false)
+
         setupUrlInput(layout)
         initToolbar(layout, viewLifecycleOwner, fragmentManager)
     }
 
+    private fun showTooltip(navBarButton: View) {
+        tooltip.contentView.tooltip.text = navBarButton.contentDescription
+        tooltip.isClippingEnabled = false
+
+        // Since the height and width are WRAP_CONTENT, we need to force the layout measure to happen before we can
+        // access the true dimensions
+        tooltip.contentView.measure(tooltip.contentView.width, tooltip.contentView.height)
```

Comment:
This should probably use the WRAP_CONTENT constants, rather than fetching them from the tooltip.